nodes/move_to_pregrasp.py

```diff
@@ -0,0 +1,952 @@
+#!/usr/bin/env python3
+
+from __future__ import annotations  # Required for type hinting a class within itself
+
+# Standard Imports
+import sys
+import threading
+from enum import Enum
+from typing import Dict, Generator, List, Optional, Tuple
+
+# Third-Party Imports
+import message_filters
+import numpy as np
+import numpy.typing as npt
+import rclpy
+import ros2_numpy
+import tf2_py as tf2
+import tf2_ros
+import yaml
+from constants import (
+    Frame,
+    Joint,
+    get_gripper_configuration,
+    get_pregrasp_wrist_configuration,
+    get_stow_configuration,
+)
+from geometry_msgs.msg import Point, Quaternion, Transform, TransformStamped, Vector3
+from rclpy.action import ActionServer, CancelResponse, GoalResponse
+from rclpy.action.server import ServerGoalHandle
+from rclpy.callback_groups import ReentrantCallbackGroup
+from rclpy.duration import Duration
+from rclpy.executors import MultiThreadedExecutor
+from rclpy.node import Node
+from sensor_msgs.msg import CameraInfo, Image, PointCloud2
+from std_msgs.msg import Header
+from stretch_ik_control import (
+    MotionGeneratorRetval,
+    StretchIKControl,
+    TerminationCriteria,
+    remaining_time,
+)
+from tf2_geometry_msgs import PoseStamped
+from tf_transformations import quaternion_about_axis, quaternion_multiply
+
+# Local Imports
+from stretch_web_teleop.action import MoveToPregrasp
+
+# TODO: Replace many of the info logs with debugs.
+
+
+class MoveToPregraspState(Enum):
+    """
+    The below states can be strung together to form a state machine that moves the robot
+    to a pregrasp position. The general principle we follow is that the robot should only
+    rotate its base and move the lift when its arm is within the base footprint of the robot
+    (i.e., the arm length is fully in and the wrist is stowed).
+    """
+
+    STOW_ARM_LENGTH_PARTIAL = (
+        -1
+    )  # Stow the arm until the gripper would collide with the base if the gripper were vertically down.
+    STOW_ARM_LENGTH_FULL = 0  # Stow the arm fully.
+    STOW_WRIST = 1
+    STOW_ARM_LIFT = 2
+    ROTATE_BASE = 3
+    HEAD_PAN = 4
+    LIFT_ARM = 5
+    MOVE_WRIST = 6
+    LENGTHEN_ARM = 7
+    TERMINAL = 8
+
+    @staticmethod
+    def get_state_machine(
```

Comment:
Do we have a standard way to set up state machines like this? I dont love using enums

Reply:
Not something you should change or anything

Reply:
Don't have a standard way of doing it. I liked your `task.py` formulation, and if we want to move a bit more generic, in the past I've really liked behavior trees.

For this node in particular, I've considered making a separate "state" class, and each of these states is an instance of the class, and each state owns the logic of creating its own motion generator (as opposed to it being done in `execute_action` below). However, I feel that's low-priority at the moment.

Reply:
Yeah I also like behavior trees

Reply:
Agree